File: Carpentry.Angular/ClientApp/src/app/common/filter-service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { forkJoin, Observable, of } from "rxjs";
import { map, tap } from "rxjs/operators";
import { HttpService } from "../common/HttpService";
import { AppFiltersDto, FilterOption } from "../settings/models";

@Injectable({
    providedIn: 'root',
})
export class FilterService extends HttpService {

    //Can I attempt to store cached values of commonly-used things?...
    private formatFilters: FilterOption[] = [];

    
    constructor(http: HttpClient) {
        super(http);
    }

    getCardSetFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetCardSetFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }

    getCardTypeFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetCardTypeFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }

    getFormatFilters(): Observable<FilterOption[]> {

        if(this.formatFilters.length === 0){
            const endpoint = 'api/Core/GetFormatFilters';
            return this.http.get<FilterOption[]>(endpoint)
                .pipe(tap(
                    data => this.formatFilters = data,
                    error => console.log(error),
                ))
        }
        else {
            return of(this.formatFilters)
        }
    }

    getManaTypeFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetManaTypeFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }
    
    getRarityFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetRarityFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }
    
    getStatusFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetStatusFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }
    
    getCardGroupFilters(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetCardGroupFilters';
        return this.http.get<FilterOption[]>(endpoint);
    }
    
    getInventorySortOptions(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetInventorySortOptions';
        return this.http.get<FilterOption[]>(endpoint);
    }

    getInventoryGroupOptions(): Observable<FilterOption[]> {
        const endpoint = 'api/Core/GetInventoryGroupOptions';
        return this.http.get<FilterOption[]>(endpoint);
    }

    getAppFilterOptions(): Observable<AppFiltersDto> {
        return forkJoin({
            sets: this.getCardSetFilters(),
            types: this.getCardTypeFilters(),
            formats: this.getFormatFilters(),
            colors: this.getManaTypeFilters(),
            groupBy: this.getInventoryGroupOptions(),
            rarities: this.getRarityFilters(),
            searchGroups: this.getCardGroupFilters(),
            sortBy: this.getInventorySortOptions(),
            statuses: this.getStatusFilters(),
        })
        .pipe(
            map(response => {
                let filters: AppFiltersDto = {
                    types: response.types,
                    sets: response.sets,
                    formats: response.formats,
                    colors: response.colors,
                    groupBy: response.groupBy,
                    rarities: response.rarities,
                    searchGroups: response.searchGroups,
                    sortBy: response.sortBy,
                    statuses: response.statuses,
                };
                return filters;
            })
        );
    }
}

File: Carpentry.Logic/FilterService.cs
﻿using Carpentry.CarpentryData;
using Carpentry.Logic.Models;
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Carpentry.Logic
{
    public interface IFilterService
    {
        Task<AppFiltersDto> GetAppCoreData();
        Task<List<FilterOption>> GetFormatFilters();
        Task<List<FilterOption>> GetRarityFilters();
        Task<List<FilterOption>> GetCardSetFilters();
        Task<List<FilterOption>> GetStatusFilters();
        List<FilterOption> GetCardTypeFilters();
        List<FilterOption> GetManaTypeFilters();
        List<FilterOption> GetCardGroupFilters();
        List<FilterOption> GetInventoryGroupOptions();
        List<FilterOption> GetInventorySortOptions();
    }

    public class FilterService : IFilterService
    {
        private readonly CarpentryDataContext _cardContext;

        public FilterService(CarpentryDataContext cardContext)
        {
            _cardContext = cardContext;
        }

        public async Task<AppFiltersDto> GetAppCoreData()
        {
            var result = new AppFiltersDto
            {
                Formats = await GetFormatFilters(),
                Rarities = await GetRarityFilters(),
                Sets = await GetCardSetFilters(),
                Statuses = await GetStatusFilters(),
                Types = GetCardTypeFilters(),
                Colors = GetManaTypeFilters(),
                SearchGroups = GetCardGroupFilters(),
                GroupBy = GetInventoryGroupOptions(),
                SortBy = GetInventorySortOptions(),
            };

            return result;
        }

        public async Task<List<FilterOption>> GetFormatFilters()
        {
            return await _cardContext.MagicFormats
                .Select(x => new FilterOption()
                {
                    Value = x.FormatId.ToString(),
                    Name = x.Name
                }).ToListAsync();
        }

        public async Task<List<FilterOption>> GetRarityFilters()
        {
            return await _cardContext.Rarities
                .Select(x => new FilterOption()
                {
                    Value = x.RarityId.ToString(),
                    Name = x.Name
                }).ToListAsync();
        }
    
        public async Task<List<FilterOption>> GetCardSetFilters()
        {
            return await _cardContext.Sets
                .Where(s => s.IsTracked)
                .OrderByDescending(s => s.ReleaseDate)
                .Select(x => new FilterOption()
                {
                    Value = x.Code.ToString(),
                    Name = x.Name
                }).ToListAsync();
        }

        public async Task<List<FilterOption>> GetStatusFilters()
        {
            return await _cardContext.CardStatuses
                .Select(s => new FilterOption()
                {
                    Value = s.CardStatusId.ToString(),
                    Name = s.Name
                }).ToListAsync();
        }

        public List<FilterOption> GetCardTypeFilters()
        {
            return new List<FilterOption>()
            {
                new FilterOption(){ Value = "Creature", Name = "Creature" },
                new FilterOption(){ Value = "Instant", Name = "Instant" },
                new FilterOption(){ Value = "Sorcery", Name = "Sorcery" },
                new FilterOption(){ Value = "Enchantment", Name = "Enchantment" },
                new FilterOption(){ Value = "Land", Name = "Land" },
                new FilterOption(){ Value = "Planeswalker", Name = "Planeswalker" },
                new FilterOption(){ Value = "Artifact", Name = "Artifact" },
                new FilterOption(){ Value = "Legendary", Name = "Legendary" },
            };
        }

        public List<FilterOption> GetManaTypeFilters()
        {
            return new List<FilterOption>()
            {
                new FilterOption() {Value = "W", Name = "White"},
                new FilterOption() {Value = "U",Name = "Blue"},
                new FilterOption() {Value = "B",Name = "Black"},
                new FilterOption() {Value = "R",Name = "Red"},
                new FilterOption() {Value = "G",Name = "Green"},
            };
        }

        public List<FilterOption> GetCardGroupFilters()
        {
            return new List<FilterOption>()
            {
                new FilterOption() { Value = "Red", Name = "Red" },
                new FilterOption() { Value = "Blue", Name = "Blue" },
                new FilterOption() { Value = "Green", Name = "Green" },
                new FilterOption() { Value = "White", Name = "White" },
                new FilterOption() { Value = "Black", Name = "Black" },
                new FilterOption() { Value = "Multicolored", Name = "Multicolored" },
                new FilterOption() { Value = "Colorless", Name = "Colorless" },
                new FilterOption() { Value = "Lands", Name = "Lands" },
                new FilterOption() { Value = "RareMythic", Name = "RareMythic" },
            };
        }

        public List<FilterOption> GetInventoryGroupOptions()
        {
            return new List<FilterOption>()
            {
                new FilterOption() { Value = "name", Name = "Name" },
                new FilterOption() { Value = "print", Name = "Print" },
                new FilterOption() { Value = "unique", Name = "Unique" },
            };
        }

        public List<FilterOption> GetInventorySortOptions()
        {
            return new List<FilterOption>()
            {
                new FilterOption() { Value = "count", Name = "Count" },
                new FilterOption() { Value = "name", Name = "Name" },
                new FilterOption() { Value = "price", Name = "Price" },
                new FilterOption() { Value = "cmc", Name = "Cmc" },
                new FilterOption() { Value = "collectorNumber", Name = "Collector Number"}
            };
        }
    }
}
